File: python/src/linkedlists/remove_n_items.py
from typing import Optional

from src.linkedlists.utils import ListNode


class Solution:
    def removeNthFromEnd(self, head: Optional[ListNode], k: int) -> Optional[ListNode]:
        dummyNode = ListNode(None, head)
        leftNode = dummyNode
        rightNode = head
        while k > 0 and rightNode:
            rightNode = rightNode.next
            k -= 1
        while rightNode:
            rightNode = rightNode.next
            leftNode = leftNode.next
        leftNode.next = leftNode.next.next
        return dummyNode.next

File: java/src/main/java/com/blind75/linkedlists/RemoveNItems.java
package com.blind75.linkedlists;

public class RemoveNItems {
    public ListNode removeNthFromEnd(ListNode node, int k) {
        var dummyNode = new ListNode(0, node);
        var leftNode = dummyNode;
        var rightNode = node;

        while (rightNode != null && k > 0) {
            rightNode = rightNode.next;
            k--;
        }

        while (rightNode != null) {
            rightNode = rightNode.next;
            leftNode = leftNode.next;
        }

        leftNode.next = leftNode.next.next;
        return dummyNode.next;
    }
}
